decentraland/unity-explorer
# Fix: scene start up failure

Fixes #2684

## Got rid of empty scenes

There was an issue

```
NullReferenceException: Object reference not set to an instance of an object.
  at ECS.SceneLifeCycle.Systems.UpdateCurrentSceneSystem.Update (System.Single t) [0x00000] in <00000000000000000000000000000000>:0 
  at ECS.Abstract.BaseUnityLoopSystem.Update (System.Single& t) [0x00000] in <00000000000000000000000000000000>:0 
```

which indicates `ISceneStateProvider` is not set for the scene facade.

Instead of finding and fixing a potential source of the issue I got rid of the concept of empty scene entirely as it's been completely redundant since we introduced `Terrain`:
- There will be no entity representing an empty scene that will lead faster querying over `SceneEntityDefinitionComponent`
- Scene readiness will be resolved immediately just like for LODs, and roads.

## Centralized handling of `TeleportIntent`

Now the single system `TeleportCharacterSystem` is responsible for the whole flow of handling `TeleportIntents`:
- Happy path (before it was only this)
- Cancellation (including Timeout), all local timeouts were removed
- Failure with exceptions
- Synchronizing with the existing queue
- Recovery from failed teleport intents

## Strict binding of loading progress with loading screen

There was no direct binding of the state of `AsyncLoadProcessReport` with the view.

Empowered `LoadingScreen` to handle all possible outcomes of the process (which is driven externally) and reflect them on the report itself, including:
- TimeOut
- Hard Cancellation
- Failure of the internal process
- Happy path 

Now, UI fully relies on the state of `AsyncLoadProcessReport` and does not do any checks of timeouts, errors, etc locally. 

## Exception-free flow is fixed up

There was a mix of exceptions and `Results` in Teleport and Start-up flows. The guidelines are [formalized](https://github.com/decentraland/unity-explorer/wiki/Architecture-Overview#exceptions-free-async-flow) and the code was refactored to follow this principle.

Before it was hard to follow the flow that can freely violate the contracts.

## Related issues

The loading process can be interrupted at any time, leaving the game in an unpredictable state. This PR does not cover this case as it's much wider. It will be addressed in the separate issue #2821


## How to QA:

- Try to break the loading process by:
 - teleporting between realms, empty parcels, and real scenes
 - adjusting the debug loading timeout (by default is 2 minutes) to low values to simulate cancellation in the middle of the loading process. I expect you will find more bugs for which you will be able to create tickets (and later we will address them as part of #2821
 - no matter what the loading screen should always correspond to the loading process: whether it's failed, canceled or succeeded, and never be stuck
 - Even if teleport has failed the user should be properly teleported back, and all the surroundings loaded properly (inc. **without moving**)

![image](https://github.com/user-attachments/assets/8e38a14b-6ae8-4873-b386-d6280d3bd26a)
